client/core/bond.go

```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+//  - create a replacement bond at lockTime-BondExpiry-pendingBuffer
+
+// pendingBuffer gives the duration in seconds prior to reaching bond expiry
+// (account, not lockTime) after which a new bond should be posted to avoid
+// account tier falling below target while the replacement bond is pending. The
+// network is a parameter because expected block times are network-dependent,
+// and we require short bond lifetimes on simnet.
+func pendingBuffer(net dex.Network) int64 {
+	switch net {
+	case dex.Mainnet: // unpredictable, so extra large to prevent falling tier
+		return 90 * 60
+	case dex.Testnet: // testnet generally has shorter block times, min diff rules, and vacant blocks
+		return 20 * 60
+	default: // Regtest and Simnet have on-demand blocks
+		return 35
+	}
+}
+
+// spendableDelay gives a high estimate in seconds of the duration required for
+// a bond to become spendable after reaching lockTime. This depends on consensus
+// rules and block times for an asset. For some assets this could be zero, while
+// for others like Bitcoin, a time-locked output becomes spendable when the
+// median of the last 11 blocks is greater than the lockTime. This function
+// returns a high value to avoid all but extremely rare (but temporary) drops in
+// tier. NOTE: to minimize bond overlap, an asset.Bonder method could provide
+// this estimate, but it is still very short relative to the entire bond
+// lifetime, which is on the order of months.
+func spendableDelay(net dex.Network) int64 {
+	// We use 3*pendingBuffer as a well-padded estimate of this duration. e.g.
+	// with mainnet, we would use a 90 minute pendingBuffer and a 270 minute
+	// spendableDelay to be robust to periods of slow blocks.
+	return 3 * pendingBuffer(net)
+}
+
+// minBondLifetime gives the minimum bond lifetime (a duration from now until
+// lockTime) that a new bond should use to prevent a tier drop. bondExpiry is in
+// seconds.
+func minBondLifetime(net dex.Network, bondExpiry int64) time.Duration {
+	lifeTimeSec := 2*(pendingBuffer(net)+bondExpiry) + spendableDelay(net) // 2*(p+E)+m
+	return time.Second * time.Duration(lifeTimeSec)
+}
+
+func (c *Core) rotateBonds(ctx context.Context) {
 	// 1. Refund bonds with passed lockTime.
 	// 2. Move bonds that are expired according to DEX bond expiry into
 	//    expiredBonds (lockTime<lockTimeThresh).
-	// 3. TODO: Add bonds to keep N bonds active, according to some optional config.
+	// 3. Add bonds to keep N bonds active, according to target tier and max
+	//    bonded amount, posting before expiry of the bond being replaced.
 	now := time.Now().Unix()
 
 	type bondID struct {
 		assetID uint32
 		coinID  []byte
 	}
 
+	bondKey := c.bondKey()
+	if bondKey != nil {
+		defer bondKey.Zero() // it's a copy
+	}
+
 	for _, dc := range c.dexConnections() {
+		// time.Now().After(time.Unix(bond.LockTime-dc.cfg.BondExpiry))
 		lockTimeThresh := now // in case dex is down, expire (to refund) when lock time is passed
```

Comment:
"refund when lock time is passed" is grouped separately because "expire" in this case means when the bond ceases to become active for tier accounting purposes, which happens some duration before lockTime, and that duration is the "expiry" duration

Reply:
"refund when lock time is passed" isn't grouped separately currently, it's currently ...
`expire (to refund) when lock time is passed`

Reply:
ah, ok you are correct!